streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -325,6 +334,12 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
         }
         // write the checkpoint file before closing, to indicate clean shutdown
         try {
+            if (checkpoint == null) {
+                assert eosEnabled;
+                checkpoint = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
+            } else {
+                assert !eosEnabled;
```

Comment:
Assertions are not enabled when we run tests, actually. They could do be, but they are not at the moment.

The other points... well, that's a larger discussion. I'll just say this: if something is worth checking during development, it's usually worth checking in production... unless the overhead is significant enough for one to have to compromise. And, in most cases, it's not.